Add a way to generate CSS custom properties with Design tokens

joelcarr opened the pull request:
This PR adds the functionality to be able to define Design tokens to be used as CSS custom properties. This is done by adding a optional `design-tokens` map in the `$gorko-config` of your design tokens and you can add `use-design-token` to your utility structure to use the CSS custom properties for your utility classes.

Example config:

```scss
/* config.scss */
$gorko-size-scale: (
  '300': $gorko-base-size * 0.8,
  '400': $gorko-base-size,
  '500': $gorko-base-size * 1.25,
  '600': $gorko-base-size * 1.6,
  '700': $gorko-base-size * 2,
  '900': $gorko-base-size * 3
);

$gorko-colors: (
  'dark': #1a1a1a,
  'light': #f3f3f3
);

$gorko-config: (
  'design-tokens': (
    'color': $gorko-colors,
    'size': $gorko-size-scale
  ),
  'bg': (
    'items': $gorko-colors,
    'output': 'standard',
    'property': 'background',
    'use-design-token': 'color'
  ),
  'text': (
    'items': $gorko-size-scale,
    'output': 'responsive',
    'property': 'font-size',
    'use-design-token': 'size'
  ),
  'breakpoints': (
    'lg': '(min-width: 62em)'
  )
);
```

Which will compile:

```css
/* compiled.css */
:root {
  --color-dark: #1a1a1a;
  --color-light: #f3f3f3;
  --size-300: 0.8rem;
  --size-400: 1rem;
  --size-500: 1.25rem;
  --size-600: 1.6rem;
  --size-700: 2rem;
  --size-900: 3rem;
}

.bg-dark {
  background: var(--color-dark);
}

.bg-light {
  background: var(--color-light);
}

.text-300 {
  font-size: var(--size-300);
}

.text-400 {
  font-size: var(--size-400);
}

.text-500 {
  font-size: var(--size-500);
}

.text-600 {
  font-size: var(--size-600);
}

.text-700 {
  font-size: var(--size-700);
}

.text-900 {
  font-size: var(--size-900);
}

@media screen and (min-width: 62em) {
  .lg\:text-300 {
    font-size: var(--size-300);
  }

  .lg\:text-400 {
    font-size: var(--size-400);
  }

  .lg\:text-500 {
    font-size: var(--size-500);
  }

  .lg\:text-600 {
    font-size: var(--size-600);
  }

  .lg\:text-700 {
    font-size: var(--size-700);
  }

  .lg\:text-900 {
    font-size: var(--size-900);
  }
}
```

Closes #8.

joelcarr committed:
Add a way to define design token custom properties

joelcarr committed:
Add design token documentation to readme.md

joelcarr committed:
Remove design tokens config from default config

Andy-set-studio commented on readme.md:
```diff
@@ -315,6 +329,43 @@ If we set the `output` to be `responsive`, with the default `breakpoints` define
 
+### `use-design-token`
+
+Optionally you can define `use-design-token` with a [`design-tokens`]() key as the value in the utility class structure. This will set the property value as the design token CSS custom property.
```
Whoops, missing link here

Andy-set-studio commented on src/generator/_generator.scss:
```diff
@@ -2,6 +2,19 @@
 
 /// Only run if there should be an output
 @if ($outputTokenCSS == true) {
+  $design-tokens:   map-get($gorko-config, 'design-tokens');
```
I think you have one of those auto-spacing things going one here, but let's keep it as a single space.

Andy-set-studio reviewed (changes requested):
I love the way this is going, Joel. Thank you.

I do have concerns that `use-design-tokens` is determining wether we use custom properties or not, which goes against the grain of a design token being a flexible system.

I do _really_ like that we are setting a design tokens map in the config and then referencing their key in the utility class definitions. I think that's very clean and will be helpful where the design tokens are set away from code.

Can we explore adding another level of control for the utility classes? Much like how we discussed in #8, some sort of flag in `$gorko-config`, like `css-vars: true` or `output-type: 'standard'|'vars'`? 

Thanks again for making a dent in this!

hankchizljaw commented:
cc @robdodson

It'll be great to see your thoughts on this too :)

joelcarr commented:
> I do have concerns that `use-design-tokens` is determining wether we use custom properties or not, which goes against the grain of a design token being a flexible system.

A righteous concern. I was worried about the naming of it and I got myself too stuck into the idea that this is for design tokens.

> I do _really_ like that we are setting a design tokens map in the config and then referencing their key in the utility class definitions. I think that's very clean and will be helpful where the design tokens are set away from code.

Thank you! Definitely need to work on the naming of these features but glad it is heading in the right direction. 
 
> Can we explore adding another level of control for the utility classes? Much like how we discussed in #8, some sort of flag in `$gorko-config`, like `css-vars: true` or `output-type: 'standard'|'vars'`?

Why of course. Would this be a flag that will create CSS variables from your utility class? For example:

```scss
/* config.scss */
$gorko-config: (
  'width': (
    'items': (
      'full': '100%',
      'half': percentage(1/2)
    ),
    'output': 'responsive',
    'css-vars': true,
    'property': 'width'
  )
)

```

```css
/* compiled.css */
:root {
  --width-full: 100%;
  --width-half: 50%;
}

.width-full {
  width: var(--width-full);
}


.width-half {
  width: var(--width-half);
}
```

robdodson commented:
This is great!

In terms of flag names, I may lean toward `css-vars` or just `variables: true`, but I don't have a strong opinion.

hankchizljaw commented:
Yeh I'd go with `css-vars` to push them away from the SCSS probably?

joelcarr commented:
Hey Folks 👋 Sorry had been away and focused on client work! This has still very much been on my mind.

I have a rough version of a utility in the `$gorko-config` having a `css-vars` property which can either be a boolean to create a variable based off the utility name or a string value which will be name of the variable you wish to use. For example:

```scss
$gorko-config: (
  'column': (
    'items': (
      'full': '100%',
      'half': percentage(1/2)
    ),
    'output': 'standard',
    'css-vars': true,
    'property': 'width'
  )
  'text': (
    'items': (
      '400': '1rem',
      '500': '1.25rem'
    ),
    'output': 'standard',
    'css-vars': 'size',
    'property': 'font-size'
  )
)
```

```css
:root {
  --column-full: 100%;
  --column-half: 50%;
}

.column-full {
  width: var(--column-full);
}

.column-half {
  width: var(--column-half);
}

:root {
  --size-400: 1rem;
  --size-500: 1.25rem;
}

.text-400 {
  font-size: var(--size-400);
}

.text-500 {
  font-size: var(--size-500);
}
```

Is this what you both had in mind? If you had a different idea I would love to see an example config of your idea.

robdodson commented:
that's interesting. My hunch is I would probably only use the `css-var: true` argument, but I don't see any immediate downside to the `css-vars: [string]` approach, except it might add a tiny bit of additional code to the implementation. @joelcarr did you run into a situation where using `css-vars: [string]` felt better than `css-vars: true`?

joelcarr commented:
The thing I am struggling to figure out is how will the utility know which
CSS variable to use? If we create a `text` utility that needs to use the
`size` CSS variable,  how do we get `css-var: true` to translate that?

On Wed, 7 Oct 2020 at 19:36, Rob Dodson <notifications@github.com> wrote:

> that's interesting. My hunch is I would probably only use the css-var:
> true argument, but I don't see any immediate downside to the css-vars:
> [string] approach, except it might add a tiny bit of additional code to
> the implementation. @joelcarr <https://github.com/joelcarr> did you run
> into a situation where using css-vars: [string] felt better than css-vars:
> true?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/hankchizljaw/gorko/pull/9#issuecomment-705120440>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG2CGG2Y7UBJK6QPSZJWFTSJSYLFANCNFSM4RSF2ZSA>
> .
>

robdodson commented:
>  If we create a `text` utility that needs to use the `size` CSS variable,  how do we get `css-var: true` to translate that?

I think I'm following... Is it an issue of reuse? Are you asking: if `pad` and `gap` both need to use the same underlying css variables, how would we express that?

hankchizljaw commented:
I think the `'css-vars': 'size'` route is nice tbh. It does have me worried about confusing someone who isn't aware of what gorko does though.

maybe something like this?

```scss
'css-vars': true,
'token-group': 'size'
```

joelcarr committed:
wip! Use bool of css-vars to set use of variables

redhwannacef commented:
Hey everyone, great work on this so far, looking forward to using it!

I thought I would add a quick suggestion. I wonder if adding the two separate variables (`css-vars`, `token-group`) may cause more confusion as it adds more failure scenarios?  Because they are tightly coupled and don't work without each other, only adding `'css-vars': true` looks like it will output invalid tokens, only adding `token-group` will be ignored, and then there adding a `token-group` that don't not exist in the tokens map will output invalid tokens too.

Maybe the previous suggestion is more intuitive: For example one property `'use-token': 'color'`, then to be safe an additional build time error message if the provided token does not exist in the `tokens` map? (Something explicit like, `x was not found in the tokens map... To use tokens please ensure it is added...`).

hariioom commented:
Any News on if and when this PR will be merged, @hankchizljaw? Would love using it...

hankchizljaw commented:
> Any News on if and when this PR will be merged, @hankchizljaw? Would love using it...

It’s still WIP at the moment 👍

andy-piccalilli commented:
Hey @joelcarr, do you mind if I pick this one up? I'm going to need it for a project :)

joelcarr commented:
Hey @andy-piccalilli 👋  I do not mind at all! Sorry I for having this just sit here, it has been on my mind but my mind out of work has been in full relax mode.

andy-piccalilli commented:
Oh no worries at all! It's so very close already and I had completely forgotten about it myself too lol. 

Thanks, @joelcarr :)

Andy-set-studio committed:
Re-work how vars are defined and interacted with

On second look, the 'design-tokens' stuff didn't sit comfortably
because the config as a whole is token-based. To clear things up
I've set it that 'tokens' is now 'css-vars' in the config and to use
them, a collection sets 'css-vars' to true, then changes the 'items'
to be a reference to the 'css-vars' config. Works nice!

Andy-set-studio committed:
Update docs

Andy-set-studio committed:
Set default config back to default

andy-piccalilli commented:
Right this is ready to go. I've kept it mostly the same but tweaked how the utility class groups work. It's much easier to define a group as a custom property now with much less repetition: https://github.com/andy-piccalilli/gorko/blob/4b93e8281f007970b3a353ebe89f762f0a6fdca1/readme.md#using-css-custom-properties

Andy-set-studio committed:
Bump and polish up docs
